README

                 Thot library and Amaya

Thot is a generic system for the development of document-centered
applications based on the concept of structured active documents. It can be
used for building interactive system as well as automatic processors.
Currently the open-source program using the Thot Library is Amaya
(http://www.w3.org/Amaya).
For more information about Amaya. see the README.amaya file
located in /usr/share/doc/amaya
and the documentation located in /usr/share/amaya/doc/WX.


redland/NEWS

                   Redland RDF Application Framework - News

2004-02-24 Redland 0.9.16 Released

   A  small  but  important  bug fix to make the installed C headers work
   correctly again.

   See the [1]0.9.16 Release Notes for the full details of the changes.

2004-01-31 Redland 0.9.15 Released

   API  CHANGE  Several  classes  were modified to use unsigned char* for
   UTF-8 encoded URIs and Literals
   Update  to  synchronise  with  [2]Raptor  version  1.2.0  (RDF/XML and
   N-Triples parser) - see [3]Raptor NEWS gaining the [4]Turtle Terse RDF
   Triple Language
   Added storage with MySQL written by [5]Morten Frederiksen
   Added storage with [6]AKT Triplestore
   Added file and URI-backed in-memory storage
   Added  fixes  to  make  Redland compile cleanly under C++ (tested with
   g++)
   Added assertion checks and reporting for function arguments
   Added method for finding statements in a context
   Added a method for listing the contexts in a graph
   Storage,  model  and  serializer  get/set feature methods now take and
   return nodes
   Improved  rdfproc  utility  to  handle multiple storage types, storage
   options, contexts and environment variables for configuration.
   Fix crash on reporting parser errors with % in the message
   Improved RDF/XML serializer to chose better output names
   Many other minor changes and fixes.

   See  the  [7]Detailed  0.9.15  Release  Notes for the full list of the
   changes.

2003-09-08 Redland 0.9.14 Released

   Update  to  synchronise  with  [8]Raptor  version  1.0.0  (RDF/XML and
   N-Triples parser) - see [9]Raptor NEWS for the changes since 0.9.12.
   Added  a  new  utility  program  rdfproc - a general RDF processor for
   Redland that allows command line use of the library.
   Made the Node class intern every node.
   Fixed the broken use of Perl UNIVERSAL::isa
   Other minor API changes

   See  the  [10]Detailed  0.9.14  Release Notes for the full list of the
   changes.

2003-08-28 Redland 0.9.13 Released

   API  CHANGE  [11]Node The class was modifed to make objects immutable,
   removing the set methods and to add new node type convienence methods.
   API addition [12]Model Added a method to sync the model to the backing
   store.
   API addition [13]Parser Added methods to allow parsing from strings.
   API   addition   [14]Statement  Added  method  to  check  for  partial
   statements.
   API addition [15]Storage Added a method to sync the backing store.
   Better error recovery - most errors return if possible.
   Added initial simple threading support (patches by Seth Ladd)
   Greatly  improved  Python API using 2.2 features (Edd Dumbill and Matt
   Biddulph)
   Updated  Perl,  Python  and  Java  APIs to always return new nodes and
   statements, removing the need to consider shared object issues.
   Updated serializing to RDF/XML to handle datatype literals
   Update  to  [16]Raptor version 0.9.12 (RDF/XML and N-Triples parser) -
   see [17]Raptor NEWS for changes since 0.9.8.
   Extensive portability and other bug fixes.

   See  the  [18]Detailed  0.9.13  Release Notes for the full list of the
   changes.

2003-02-13 Redland 0.9.12 Released

   FORMAT CHANGE The persistent storage format was changed to support RDF
   typed  literals  and  this  will  require  an  upgrade of any existing
   Berkeley/Sleepycat  DB  stores created by Redland 0.9.11 or earlier. A
   utility  redland-db-upgrade  is provided that creates an updated store
   from an existing one.
   API  CHANGE  [19]Iterator  and  [20]Stream  classes. The next/get_next
   method  is  split into get_object always returning a shared object and
   next  methods. get_context method added. (get_object is called current
   in higher level language APIs)
   API  CHANGE  [21]Model  class.  Method  add_statement  no longer takes
   ownership   of  the  passed  in  statement.  The  caller  now  retains
   ownership.
   API  CHANGE [22]Model class. Method add_statements no longer frees the
   passed in librdf_stream; the caller must free it now.
   API  CHANGE  [23]Node  and  [24]Model  classes.  Added  RDF  Datatyped
   Literals. Other literal is_wf_xml constructors and methods deprecated.
   API  CHANGE  All  literals methods and constructors have no mention of
   the never-used or supported XML Space argument which is not part of an
   RDF literal.
   Added  a  script [25]utils/update-api-0912.pl to help automate the API
   changes  as  far  as  possible  or  warn  about  those  that cannot be
   automatically updated.
   Added  contexts  - Statements added to [26]Model (when using a Storage
   that supports contexts) can take an optional [27]Node, returned by the
   new get_context method on [28]Iterator and [29]Stream classes.
   Many  portability  and compiling fixes (OSX, gcc 3.x, FreeBSD, non-gcc
   compiler)
   [30]Java   API   classes.  Add  finished()  method  replacing  useless
   finalize().
   Use Raptor message callbacks to get parser warnings and errors

davlib/davlib.c
/*  --------------------------------------------------------
 ** 
 ** File: davlib.c - WebDAV module
 **
 ** This file implements the WebDAV functions for Amaya editor.
 **
 ** Author : Manuele Kirsch Pinheiro
 ** Email: Manuele.Kirsch_Pinheiro@inrialpes.fr / manuele@inf.ufrgs.br 
 ** Project CEMT - II/UFRGS - Opera Group/INRIA
 ** Financed by CNPq(Brazil)/INRIA(France) Cooperation
 **
 ** Date : April / 2002
 **
 ** $Id: davlib.c,v 1.18 2008/05/07 13:49:07 kia Exp $
 ** $Date: 2008/05/07 13:49:07 $
 ** $Log: davlib.c,v $
 ** Revision 1.18  2008/05/07 13:49:07  kia
 ** char* => const char*
 ** (Regis patches + many many others)
 **
 ** Revision 1.17  2005/06/23 15:00:48  cvs
 ** fix 2 memory leak.
 ** (I hope it can help for MacOSX crash on http://www.w3.org/People/all ?)
 ** S. GULLY
 **
 ** Revision 1.16  2005/06/21 16:02:06  cvs
 ** Bug fix: a double free in davlib code caused the
 ** wx version built with dav option to crash.
 ** LC + SG
 **
 ** Revision 1.15  2005/06/01 13:48:48  cvs
 ** Fixing some memory leaks.
 ** IV + SG
 **
 ** Revision 1.14  2005/02/04 12:18:24  vatton
 ** Improve the transformation into a table
 ** + typo in the WebDAV documentation.
 ** Irene
 **
 ** Revision 1.13  2003/12/16 12:10:08  cvs
 ** Rename flags _WINDOWS by _WINGUI
 ** + the flag _WINDOWS must be used only for windows specific code.
 ** + the flag _WINGUI must be used only for old windows gui stuff.
 **
 ** S. GULLY
 **
 ** Revision 1.12  2003/11/19 12:33:16  gully
 ** Compilation fix (webdav support)
 **
 ** S. GULLY
 **
 ** Revision 1.11  2003/05/19 10:24:54  vatton
 ** Internationalization of the MakeID dialogue.
 ** Irene
 **
 ** Revision 1.10  2002/07/01 10:34:16  kahan
 ** JK: Enabling/Disabling DAV support by means of the new DAV_Enable
 ** registry entry.
 ** Removed the DAV menu from the standard profile.
 **
 ** Revision 1.9  2002/06/13 16:10:13  kirschpi
 ** New dialogue "WebDAV Preferences"
 ** Corrections due last commit by JK
 ** Manuele
 **
 ** Revision 1.8  2002/06/13 13:40:32  kahan
 ** JK: Changed MAX_LINE to DAV_MAX_LINE. MAX_LINE is a reserved macro and
 ** the code was generating a warning.
 **
 ** Revision 1.7  2002/06/11 17:34:52  kirschpi
 ** improving commentaries and format (for WebDAV code).
 ** update DAV Resources list automatically
 ** Manuele
 **
 ** Revision 1.6  2002/06/06 17:10:45  kirschpi
 ** Breaking the user messages in three lines
 ** Fixing some code format problems
 ** Fixing DAVLockIndicator, when Lock discovery is disabled.
 ** Fixing unecessary memory allocations in FilterMultiStatus_handler
 ** and FilterLocked_handler.
 ** Manuele
 **
 ** Revision 1.5  2002/06/05 16:46:06  kirschpi
 ** Applying Amaya code format.
 ** Modifying some dialogs (looking for a better windows presentation)
 ** Adding a DAVResource list, a list of resources (specially collections),
 ** where we should do a Lock discovery.
 ** Manuele
 **
 ** Revision 1.4  2002/06/04 14:51:41  kirschpi
 ** Fixing bug on InitDAV.
 ** Manuele
 **
 ** Revision 1.3  2002/06/03 14:37:42  kirschpi
 ** The name of some public functions have been changed to avoid conflic with
 ** other libraries.
 **
 ** Revision 1.2  2002/05/31 17:59:19  kirschpi
 ** Functions to give to user some informations about active locks
 ** (a basic awareness support) when the user load or exit a document.
 **
 * -------------------------------------------------------- 
 */

#include <stdlib.h>
#include <stdio.h>

#define THOT_EXPORT extern

#define WEBDAV_EXPORT

#include "davlib.h"

#include "davlib_f.h"
#include "query_f.h"
#include "init_f.h"
#include "AHTURLTools_f.h"

#include "davlibCommon_f.h"
#include "davlibRequests_f.h"
#include "davlibUI_f.h"


/* ********************************************************************* *
 *                               INIT FUNCTION                           *
 * ********************************************************************* */


/*----------------------------------------------------------------------
   InitDAV - initialization function. Inits DAVHome variable pointing to
             home dir, where the lock base will be stored.   
  ----------------------------------------------------------------------*/
void InitDAV (void) 
{
    char *ptr = NULL;
    BOOL modified = FALSE;
    const char *fqdn = NULL;
    char *email = NULL;

    /* DAV support is enabled */
    /*TtaGetEnvBoolean ("DAV_ENABLE", &DAVLibEnable);
    if (!DAVLibEnable)
    return;*/
    DAVLibEnable = TRUE;
    /* ******************** DAVHome ********************* */
    /* try to get default home (.amaya) dir from APP_HOME */
    ptr = TtaGetEnvString ("APP_HOME");
    sprintf (DAVHome,"%s%c",(ptr)?ptr:TempFileDirectory,DIR_SEP);

#ifdef DEBUG_DAV
    fprintf (stderr,"InitDAV..... DAVHome is %s\n",DAVHome);
#endif   

    
    /* **************** DAVFullHostName ***************** */
    fqdn = DAVFQDN();
    sprintf (DAVFullHostName,"%s",fqdn);
    
#ifdef DEBUG_DAV
    fprintf (stderr,"InitDAV..... DAVFullHostName is %s\n",DAVFullHostName);
#endif   

    
    /***** DAVUserURL, DAVLockScope, DAVDepth, DAVTimeout *****/

    /* getting user URL: if no DAV_USER_URL element defined in 
     * thot.rc, get the user email */
    ptr = NULL; 
    ptr = TtaGetEnvString (DAV_USER_URL);
    if (ptr && (*ptr))   
        sprintf (DAVUserURL,ptr);
    else 
     {  
        email = DAVDefaultEmail();
        sprintf (DAVUserURL,"%s%s",(email)?"mailto:":"http://", (email)?email:fqdn);

        /* save user URL in thot env */
        TtaSetEnvString (DAV_USER_URL,DAVUserURL,TRUE);
        modified = TRUE;
     }
    
    /* getting lock scope. If there is an entry DAV_LOCK_SCOPE in
     * thot.rc and it is valid, use it. Otherwise, assume "exclusive" scope */
    ptr = NULL; 
    ptr = TtaGetEnvString (DAV_LOCK_SCOPE);
    if (ptr && (*ptr) && 
       (!strcmp(ptr,"exclusive") || !strcmp (ptr,"shared")))
        sprintf (DAVLockScope,ptr);
    else 
     { 
        sprintf (DAVLockScope,"exclusive");
        
        /*save lock scope in thot env*/
        TtaSetEnvString (DAV_LOCK_SCOPE,DAVLockScope,TRUE);
        modified = TRUE;
     }

    
    /* getting depth option. If there is an entry DAV_DEPTH and
     * it is valid, use it/ Otherwise, assuming 0 
     * Note: for LOCK request, only the values 0 and infinity are accepted*/
    ptr = NULL;  
    ptr = TtaGetEnvString (DAV_DEPTH);
    if (ptr && (*ptr) &&\
       (!strcmp (ptr,"0") || !strcmp(ptr,"infinity")))
        sprintf (DAVDepth,ptr); 
    else 
     {
        sprintf (DAVDepth,"0");
        
        /* save depth in thot env */
        TtaSetEnvString (DAV_DEPTH,DAVDepth,TRUE);
        modified = TRUE;
     }

    
    /* getting timeout option. If no timeout,assuming Infinite */
    ptr = NULL; 
    ptr = TtaGetEnvString (DAV_TIMEOUT);
    if (ptr && (*ptr))
        sprintf (DAVTimeout,ptr);
    else
     { 
        sprintf (DAVTimeout,"Infinite");

        /* save timeout in thot env */
        TtaSetEnvString (DAV_TIMEOUT,DAVTimeout,TRUE);
        modified = TRUE;
     }

    
    /************* DAVAwareness, DAVAwarenessOnExit ************ */
    /* getting general option about awareness information */
    ptr = NULL;
    DAVAwareness = NO;
    ptr = TtaGetEnvString (DAV_AWARENESS);
    if (ptr && (*ptr)) 
     {
        if (!strcasecomp (ptr,"yes"))
            DAVAwareness = YES; 
     }
    else 
     {
        TtaSetEnvString (DAV_AWARENESS,"no",TRUE);
        modified = TRUE;
     }
    
    /* getting option about awareness information when exiting a resouce */
    ptr = NULL;
    DAVAwarenessExit = NO;
    ptr = TtaGetEnvString (DAV_AWARENESS_EXIT);
    if (ptr && (*ptr)) 
     {
        if (!strcasecomp (ptr,"yes"))
            DAVAwarenessExit = YES; 
     }    
    else
     {
        TtaSetEnvString (DAV_AWARENESS_EXIT,"no",TRUE);
        modified = TRUE;
     }
    
    
    /************************** DAVResources ******************** */
    ptr = NULL;
    ptr = TtaGetEnvString (DAV_URLS);
    if (ptr && (*ptr))
        sprintf (DAVResources,ptr);
    else
     { 
        DAVResources[0]='\0';

        /* save timeout in thot env */
        TtaSetEnvString (DAV_URLS,DAVResources,TRUE);
        /*modified = TRUE;*/
     }

   

    /* *********************** SAVING REGISTRY ***************** */
    if (modified) TtaSaveAppRegistry();

    
#ifdef DEBUG_DAV
    fprintf (stderr,"InitDAV..... WebDAV User's preferences are \n");
    fprintf (stderr,"\tuser url: %s\n\tlock scope: %s\n",DAVUserURL,DAVLockScope); 
    fprintf (stderr,"\tdepth: %s\n\ttimeout: %s\n",DAVDepth,DAVTimeout); 
    fprintf (stderr,"\tAwareness: %s\n",(DAVAwareness)?"yes":"no"); 
    fprintf (stderr,"\tAwareness on exit: %s\n",(DAVAwarenessExit)?"yes":"no"); 
    fprintf (stderr,"\tResources: %s\n",DAVResources); 

    /*HTSetTraceMessageMask("pl");*/
#endif


    /* *********************** SETING FILTERS *************** */
    HTNet_addAfter (FilterFailedDependency_handler,"http://*", NULL, 
                                           HT_FAILED_DEPENDENCY,HT_FILTER_MIDDLE);

    HTNet_addAfter (FilterLocked_handler,"http://*", NULL, 
                                           HT_LOCKED,HT_FILTER_MIDDLE);

    HTNet_addAfter (FilterMultiStatus_handler,"http://*", NULL, 
                                           HT_MULTI_STATUS,HT_FILTER_MIDDLE);


    /* *********************** SETING MENUS ***************** */
    /* set lock indicatior toggle. initial state: FALSE */
    DAVLockIndicatorState = FALSE; 
    DAVSetLockIndicator(CurrentDocument);
    
}


/* ********************************************************************* *
 *                      CLOSE DOCUMENT FUNCTION                          *
 * ********************************************************************* */


/*----------------------------------------------------------------------
   DAVFreeLock - ask to user if he/she wants to free pending locks 
                 when closing the session.   
   
  NEED: deal with last document when exiting the application.
   
        Now, when exiting the application, if the document is locked
        by the user (the lock information must be in the local base),
        this function will ask whether the user wants to unlock it.
        If user agrees, an UNLOCK request will be sent. But, under
        Windows machines, this request will be killed when the application
        exit, and no unlock will be done.
                      
  ----------------------------------------------------------------------*/
void DAVFreeLock (Document docid) 
{
    BOOL ok = NO;
    char *lockinfo;
    char *relURI, *absURI, *ptr;
    char label1[DAV_LINE_MAX], label2[DAV_LINE_MAX];

    if (!DAVLibEnable)
      return;

    lockinfo = relURI = absURI = ptr = NULL;
    
    /* if user doesn't want awareness info, neither
     * awareness info on exit, return
     */ 
    if (!(DAVAwareness && DAVAwarenessExit))
        return;
    
#ifdef DEBUG_DAV
    fprintf (stderr,"Closing document %s\n",DocumentURLs[docid]);
#endif

    /* separing URI into hostname and relative parts */
    ok = separateUri (DocumentURLs[docid], DAVFullHostName, &absURI, &relURI);
    
    if (ok && absURI && relURI) 
     {
        /* if there is a lock info in the local base,
         * user has a lock for this resource.
         */
        lockinfo = DAVFindLockToken (absURI, relURI);
        if (lockinfo && *lockinfo) 
         {
            sprintf (label1,TtaGetMessage(AMAYA,AM_LOCKED),DocumentURLs[docid]);
            ptr = DAVBreakString (label1);
            sprintf (label2,TtaGetMessage(AMAYA,AM_UNLOCK_DOCUMENT));
            
            if (DAVConfirmDialog (docid,label1,(ptr)?ptr:(char *)" ", label2)) 
             {
                AHTDAVContext *new_davctx = GetUnlockInfo (docid);
                if (new_davctx) 
                 {
                    new_davctx->showIt = NO;    
                    DoUnlockRequest (docid, new_davctx);                    
                 }    
             }
         } /* lockinfo */
     }
}



/* ********************************************************************* *
 *                           REGISTRY OPERATION                          *
 * ********************************************************************* */


/*----------------------------------------------------------------------
   DAVSaveRegistry - save DAV information in the registry

  ----------------------------------------------------------------------*/
void DAVSaveRegistry (void)
{
  if (!DAVLibEnable)
    return;

    /***** DAVUserURL, DAVLockScope, DAVDepth, DAVTimeout *****/
    TtaSetEnvString (DAV_USER_URL,DAVUserURL,TRUE);
    TtaSetEnvString (DAV_LOCK_SCOPE,DAVLockScope,TRUE);
    TtaSetEnvString (DAV_DEPTH,DAVDepth,TRUE);
    TtaSetEnvString (DAV_TIMEOUT,DAVTimeout,TRUE);
    
    /************* DAVAwareness, DAVAwarenessOnExit ************ */
    if (DAVAwareness == YES) 
        TtaSetEnvString (DAV_AWARENESS, "yes",TRUE);
    else 
        TtaSetEnvString (DAV_AWARENESS, "no",TRUE);
    
    if (DAVAwarenessExit == YES)
        TtaSetEnvString (DAV_AWARENESS_EXIT, "yes",TRUE);
    else
        TtaSetEnvString (DAV_AWARENESS_EXIT, "no",TRUE);
    
    
    /************************** DAVResources ******************** */
    TtaSetEnvString (DAV_URLS,DAVResources,TRUE);


    /* *********************** SAVING REGISTRY ***************** */
    TtaSaveAppRegistry();

    
#ifdef DEBUG_DAV
    fprintf (stderr,"DAVSaveRegistry..... WebDAV User's preferences are \n");
    fprintf (stderr,"\tuser url: %s\n\tlock scope: %s\n",DAVUserURL,DAVLockScope); 
    fprintf (stderr,"\tdepth: %s\n\ttimeout: %s\n",DAVDepth,DAVTimeout); 
    fprintf (stderr,"\tAwareness: %s\n",(DAVAwareness)?"yes":"no"); 
    fprintf (stderr,"\tAwareness on exit: %s\n",(DAVAwarenessExit)?"yes":"no"); 
    fprintf (stderr,"\tResources: %s\n",DAVResources); 
#endif

}



/* ********************************************************************* *
 *                               MENU ACTIONS                            *
 * ********************************************************************* */


/*----------------------------------------------------------------------
   DAVLockDocument - Lock the document URL 

  ----------------------------------------------------------------------*/
void DAVLockDocument (Document document, View view)
{
    AHTDAVContext *davctx = NULL;

    if (!DAVLibEnable)
      return;

#ifdef DEBUG_DAV    
    fprintf (stderr,"DAVLockDocument..... Locking document %s\n",DocumentURLs[document]);
    fprintf (stderr,"DAVLockDocument..... creating the dav context object\n");
#endif
    
    /* getting DAV context object */ 
    davctx = GetLockInfo (document);
    if (!davctx) 
     {            
        DAVDisplayMessage (TtaGetMessage (AMAYA, AM_DAV_UNSUPPORTED_PROTOCOL), 
                           DocumentURLs[document]);
     }
    else 
     {
#ifdef DEBUG_DAV
        char *url = DocumentURLs[document];
        fprintf (stderr,"\tabsolute: %s\n\trelative: %s\n",davctx->absoluteURI,davctx->relativeURI);
        fprintf (stderr,"\tdepth: %s\n",HTDAV_DepthHeader(davctx->davheaders));
        fprintf (stderr,"\ttimeout: %s\n",HTDAV_TimeoutHeader(davctx->davheaders));
        fprintf (stderr,"\turl: %s\n",url);
        fprintf (stderr,"\tLock XML body: %s\n",(davctx->xmlbody)?"YES":"NO");
#endif 

        /* executing the request */
        if (!DoLockRequest (document,davctx)) 
            DAVDisplayMessage (TtaGetMessage (AMAYA, AM_LOCK_FAILED), NULL);

     }
        
}



/*----------------------------------------------------------------------
   DAVUlnlocDocument - Unlock document URL

  ----------------------------------------------------------------------*/
void DAVUnlockDocument (Document document, View view) 
{
    AHTDAVContext *davctx = NULL;

    if (!DAVLibEnable)
      return;
    
#ifdef DEBUG_DAV    
    fprintf (stderr,"DAVUnlockDocument..... Unlocking document %s\n",DocumentURLs[document]);
    fprintf (stderr,"DAVUnlockDocument..... creating the dav context object\n");
#endif
  
    davctx = GetUnlockInfo(document);
    if (davctx) 
     {
#ifdef DEBUG_DAV
        fprintf (stderr,"DAVUnlockDocument..... DAV context object:\n");
        fprintf (stderr,"\tabsolute: %s\n\trelative: %s\n",davctx->absoluteURI,davctx->relativeURI);
        fprintf (stderr,"\tdepth: %s\n",HTDAV_DepthHeader(davctx->davheaders));
        fprintf (stderr,"\tlock-token: %s\n",HTDAV_LockTokenHeader(davctx->davheaders));
#endif 
        
        /* do the request */       
        if (!DoUnlockRequest (document,davctx)) 
            DAVDisplayMessage (TtaGetMessage (AMAYA, AM_UNLOCK_FAILED), NULL);
        
        
     }
    else 
     {
        /* *** Should we give to the user the option to unlock a document anyway? YES *** */
        ForceUnlockRequest (document);
     }
}



/*----------------------------------------------------------------------
   DAVProfindDocument - do a Propfind request in the document URL 
   
  ----------------------------------------------------------------------*/
void DAVProfindDocument (Document document, View view) 
{
    AHTDAVContext *davctx= NULL;

    if (!DAVLibEnable)
      return;

#ifdef DEBUG_DAV
    fprintf (stderr,"DAVPropfindDocument.... Profind document %s\n",DocumentURLs[document]);
    fprintf (stderr,"DAVPropfindDocument..... creating the dav context object\n");    
#endif    
    
    davctx = GetPropfindInfo(document);
    if (davctx) 
     {
#ifdef DEBUG_DAV
        fprintf (stderr,"DAVPropfindDocument..... DAV context object:\n");
        fprintf (stderr,"\tabsolute: %s\n\trelative: %s\n",davctx->absoluteURI,davctx->relativeURI);
        fprintf (stderr,"\tdepth: %s\n",HTDAV_DepthHeader(davctx->davheaders));
        fprintf (stderr,"\txml body: %s\n",(davctx->xmlbody));    
#endif
        
        /* YES, show the results*/
        davctx->showIt = YES;

        if (!DoPropfindRequest (document,davctx,FilterPropfind_handler,NULL)) 
            DAVDisplayMessage (TtaGetMessage (AMAYA,AM_PROPFIND_FAILED), NULL);
                
     }
    else 
     {
#ifndef _WINGUI            
        DAVDisplayMessage (TtaGetMessage (AMAYA, AM_DAV_UNSUPPORTED_PROTOCOL), 
                           DocumentURLs[document]);                        
#endif
        DAVLockIndicatorState = FALSE;
     }   

}



/*----------------------------------------------------------------------
   DAVCopyLockInfo - do a Propfind request in the document URL 
   to discover and copy (to the local base) the lock information
  ----------------------------------------------------------------------*/
void DAVCopyLockInfo (Document document, View view) 
{
    AHTDAVContext *davctx= NULL;
    BOOL ok = NO;

    if (!DAVLibEnable)
      return;

#ifdef DEBUG_DAV
    fprintf (stderr,"DAVCopyLockInfo.... Profind document %s\n",DocumentURLs[document]);
    fprintf (stderr,"DAVCopyLockInfo..... creating the dav context object\n");    
#endif    
    
    davctx = GetPropfindInfo(document);
    if (davctx) 
     {
#ifdef DEBUG_DAV
        fprintf (stderr,"DAVPropfindDocument..... DAV context object:\n");
        fprintf (stderr,"\tabsolute: %s\n\trelative: %s\n",davctx->absoluteURI,davctx->relativeURI);
        fprintf (stderr,"\tdepth: %s\n",HTDAV_DepthHeader(davctx->davheaders));
        fprintf (stderr,"\txml body: %s\n",(davctx->xmlbody));    
#endif
        
        /* YES, show the results*/
        davctx->showIt = YES;

        if (createPropfindBody (YES,davctx->xmlbody,DAV_XML_LEN)) 
            ok = DoPropfindRequest (document,davctx,FilterCopyLockInfo_handler,NULL);
        
        
        if (!ok)
            DAVDisplayMessage (TtaGetMessage (AMAYA,AM_PROPFIND_FAILED), NULL);
        
     }
    else
     {
#ifndef _WINGUI            
        DAVDisplayMessage (TtaGetMessage (AMAYA, AM_DAV_UNSUPPORTED_PROTOCOL), 
                           DocumentURLs[document]);                        
#endif
        DAVLockIndicatorState = FALSE;
     }        
}



/*----------------------------------------------------------------------
   DAVLockIndicator: Manipulates Lock indicator
  ----------------------------------------------------------------------*/
void DAVLockIndicator (Document document, View view) 
{
        
    /* if lock indicator is TRUE, unlock the document
     * if it's FALSE, lock the document */
    if (DAVLockIndicatorState) 
        DAVUnlockDocument (document,view);    
    else  
        DAVLockDocument (document,view);    

}



/*----------------------------------------------------------------------
   DAVPreferences: shows a dialogue with WebDAV user's preferences
  ----------------------------------------------------------------------*/
void DAVPreferences (Document document, View view) 
{
    if (!DAVLibEnable)
      return;

    DAVShowPreferencesDlg (document);
}
